File: 904/solution.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        map = {}
        start = 0
        soln = 0
        for idx, tree in enumerate(fruits):
            map[tree] = map.get(tree, 0) + 1
            while len(map) > 2:
                map[fruits[start]] -= 1
                if map[fruits[start]] == 0:
                    del map[fruits[start]]
                start += 1
            soln = max(soln, idx - start + 1)
        return soln

File: 904/Solution.java
class Solution {
    public int totalFruit(int[] fruits) {
        Map<Integer, Integer> m = new HashMap<>();
        int start = 0;
        int soln = 0;
        for (int i = 0; i < fruits.length; i++) {
            m.put(fruits[i], m.getOrDefault(fruits[i], 0) + 1);
            while (m.size() > 2) {
                int f_start = fruits[start];
                m.put(f_start, m.get(f_start) - 1);
                if (m.get(f_start) == 0) {
                    m.remove(f_start);
                }
                start++;
            }
            soln = Math.max(soln, i - start + 1);
        }
        return soln;
    }
}